source/fe/fe_simplex_p.cc

```diff
@@ -756,40 +761,56 @@ std::vector<std::pair<unsigned int, unsigned int>>
 FE_SimplexP<dim, spacedim>::hp_line_dof_identities(
   const FiniteElement<dim, spacedim> &fe_other) const
 {
-  AssertDimension(dim, 2);
-
-  if (const FE_SimplexP<dim, spacedim> *fe_p_other =
+  if (const auto *fe_p_other =
         dynamic_cast<const FE_SimplexP<dim, spacedim> *>(&fe_other))
     {
-      // dofs are located along lines, so two dofs are identical if they are
-      // located at identical positions.
-      // Therefore, read the points in unit_support_points for the
-      // first coordinate direction. For FE_SimplexP, they are currently
-      // hard-coded and we iterate over points on the first line which begin
-      // after the 3 vertex points in the complete list of unit support points
+      std::vector<std::pair<unsigned int, unsigned int>> identities;
+
+      for (unsigned int i = 0; i < this->degree - 1; ++i)
+        for (unsigned int j = 0; j < fe_p_other->degree - 1; ++j)
+          if (std::fabs(this->unit_support_points[i + this->reference_cell()
+                                                        .n_vertices()][0] -
+                        fe_p_other->unit_support_points
+                          [i + this->reference_cell().n_vertices()][0]) < 1e-14)
+            identities.emplace_back(i, j);
 
+      return identities;
+    }
+  else if (const auto *fe_p_other =
+             dynamic_cast<const FE_PyramidP<dim, spacedim> *>(&fe_other))
+    {
       std::vector<std::pair<unsigned int, unsigned int>> identities;
 
       for (unsigned int i = 0; i < this->degree - 1; ++i)
         for (unsigned int j = 0; j < fe_p_other->degree - 1; ++j)
-          if (std::fabs(this->unit_support_points[i + 3][0] -
-                        fe_p_other->unit_support_points[i + 3][0]) < 1e-14)
+          if (std::fabs(this->unit_support_points[i + this->reference_cell()
+                                                        .n_vertices()][0] -
+                        fe_p_other->get_unit_support_points()
+                          [i + this->reference_cell().n_vertices()][1]) < 1e-14)
             identities.emplace_back(i, j);
-          else
-            {
-              // If nodes are not located in the same place, we have to
-              // interpolate. This is then not handled through the
-              // current function, but via interpolation matrices that
-              // result in constraints, rather than identities. Since
-              // that happens in a different function, there is nothing
-              // for us to do here.
-            }
```

Comment:
I added this in #14444 because I was wondering how this is working if the `if` above is not satisfied. I think the comment is still relevant and if you don't mind too much, please leave it here and duplicate it in the `FE_SimplexP` case above.